### ISSUE TITLE
Implement Partial Convolutions

### ISSUE THREAD

**Issue by rmholmes** [PaulSpence/OM2-emulator#11]
Ryan will copy across PartialConv2D from the ML downscaling code and try to get it working.

**Comment by rmholmes** [comment]
```
class SRDCN(nn.Module):
    """Mask-aware super-resolution convolutional network.

    Architecture:
      - Takes bilinear + nearest-neighbour high-res as input, then three masked convolutions:
      - PartialConv2d(num_features -> num_hidden, 7x7) + ReLU
      - PartialConv2d(num_hidden -> num_hidden, 5x5) + ReLU
      - PartialConv2d(num_hidden -> num_responses, 1x1)
    """

    def __init__(self, num_hidden=64, num_responses=1, num_features=1):
        super().__init__()
        self.conv1 = PartialConv2d(num_features, num_hidden, kernel_size=7, padding=3)
        self.conv2 = PartialConv2d(num_hidden, num_hidden, kernel_size=5, padding=2)
        self.conv3 = PartialConv2d(num_hidden, num_responses, kernel_size=1, padding=0)

    def forward(self, x, mask):
        # x: (B, C, H_hr, W_hr)
        x = F.relu(self.conv1(x, mask))
        x = F.relu(self.conv2(x, mask))
        x = self.conv3(x, mask)
        return x

class PartialConv2d(nn.Module):
    """
    Mask-aware (partial) convolution

    This is a simple implementation of the partial convolution from 

    Image Inpainting for Irregular Holes Using Partial Convolutions, Liu et al. 2018, NVIDIA Research

    It has no in-filling (mask is constant) and no channel-dependent mask (always uses the land-sea mask).

    Note: Here we enforce padding_mode='replicate' in order to fill 
    outwards land (i.e. assume land if point near boundary is land)
    or ocean (i.e. assume ocean if point near boundary is ocean, with 
    that points value).

    Note 2: 
    
    """
    
    def __init__(self, in_ch, out_ch, kernel_size=3, padding=1,):
        super().__init__()
        self.conv = nn.Conv2d(in_ch, out_ch, kernel_size=kernel_size, padding=padding, padding_mode='replicate') 
        
        # Mask convolution: 1->1, same geometry, no learnable params
        self.mask_conv = nn.Conv2d(1, 1, kernel_size=kernel_size, padding=padding, padding_mode='replicate', bias=False)
        self.mask_conv.weight.data[:] = 1.0
        self.mask_conv.requires_grad_(False)

    def forward(self, x, mask):
    
        # mask is 1 for ocean, 0 for land
        x_masked = x * mask
        
        out = self.conv(x_masked)
            
        # compute valid-pixel ratio in each window
        with torch.no_grad():
            mask_sum = self.mask_conv(mask)
            
        # divide by mask_sum where valid, otherwise set to zero
        kernel_area = self.conv.kernel_size[0] * self.conv.kernel_size[1]
        eps = 1e-6

        out = torch.where(mask_sum > 0, 
                          out * (kernel_area / (mask_sum + eps)), 
                          torch.zeros_like(out)
                         )
            
        return out
```

**Comment by rmholmes** [comment]
Implemented in https://github.com/PaulSpence/OM2-emulator/tree/Pconv It required changing the way we were doing the decoding - rather than using `ConvTranspose` we use bilinear upsampling and then a stride-less convolution (i.e. maintains image size).

I'm not sure it is working. In particular, there seems to be some issue with normalisation - maybe we need to fix that first?
